### ISSUE TITLE
LinkedListChannel.onSend hangs

### ISSUE THREAD

**Issue by fvasco** [Kotlin/kotlinx.coroutines#1054]
We have diagnosed a locked thread in production, the cause is due the follow code:

```kotlin
select<Unit> {
    unlimitedChannel.onSend(message) {} // <-- block here
    deferred.onJoin {}
    async.onAwait { throw it }
}
```

We get this issue using Kotlin 1.3.21, coroutines 1.1.1 on JVM 12 (maybe we got this error again using JVM 11.0.2) , AWS EC3 `c3.large` machine.

Stack trace is

```
"DefaultDispatcher-worker-2" #339 daemon prio=5 os_prio=0 cpu=9258803.68ms elapsed=10969.66s tid=0x00007feb5803a800 nid=0x169 runnable  [0x00007feb21f06000]
   java.lang.Thread.State: RUNNABLE
        at kotlinx.coroutines.selects.SelectBuilderImpl$AtomicSelectOp.prepare(Select.kt:366)
        at kotlinx.coroutines.internal.AtomicOp.perform(Atomic.kt:59)
        at kotlinx.coroutines.selects.SelectBuilderImpl.performAtomicTrySelect(Select.kt:352)
        at kotlinx.coroutines.channels.LinkedListChannel.offerSelectInternal(LinkedListChannel.kt:50)
        at kotlinx.coroutines.channels.AbstractSendChannel.registerSelectSend(AbstractChannel.kt:415)
        at kotlinx.coroutines.channels.AbstractSendChannel.access$registerSelectSend(AbstractChannel.kt:19)
        at kotlinx.coroutines.channels.AbstractSendChannel$onSend$1.registerSelectClause2(AbstractChannel.kt:398)
        at kotlinx.coroutines.selects.SelectBuilderImpl.invoke(Select.kt:407)
```

The blocked thread has an excessive CPU usage.

**Comment by qwwdfsad** [comment]
Could you please provide a reproducer?

**Comment by fvasco** [comment]
Hi @qwwdfsad 
Thank you for quick response,
I got two more stack trace, both are:

```
"DefaultDispatcher-worker-2" #339 daemon prio=5 os_prio=0 cpu=8027979.33ms elapsed=9726.60s tid=0x00007feb5803a800 nid=0x169 runnable  [0x00007feb21f06000]
   java.lang.Thread.State: RUNNABLE
	at kotlinx.coroutines.internal.AtomicOp.perform(Atomic.kt:62)
	at kotlinx.coroutines.selects.SelectBuilderImpl.performAtomicTrySelect(Select.kt:352)
	at kotlinx.coroutines.channels.LinkedListChannel.offerSelectInternal(LinkedListChannel.kt:50)
	at kotlinx.coroutines.channels.AbstractSendChannel.registerSelectSend(AbstractChannel.kt:415)
	at kotlinx.coroutines.channels.AbstractSendChannel.access$registerSelectSend(AbstractChannel.kt:19)
	at kotlinx.coroutines.channels.AbstractSendChannel$onSend$1.registerSelectClause2(AbstractChannel.kt:398)
	at kotlinx.coroutines.selects.SelectBuilderImpl.invoke(Select.kt:407)
```

The `cpu` time is constant increasing, it is really larger than others.
The procedure performs this task iterating on a map of 300 elements, we have two servers: one hangs while other works fine.

At the moment I am really busy to find a workaround, I can't provide a test case, yet.

**Comment by fvasco** [comment]
Hi @qwwdfsad 
I made some tests on unlimited channel and it looks ok, I cannot reproduce the issue.
I will look into this problem deeper (I hope this is not a library issue), I will keep you up to date.

**Comment by fvasco** [comment]
We made some more tests, the issue happens sometimes but I am not able to provide you a way to reproduce it, sorry.

We (hopefully) fixed the issue with small changes, mainly our old code was:

```kotlin
launch(start = CoroutineStart.UNDISPATCHED) {
            select<Unit> {
                messageChannel.onSend(clientMessage) {} // unlimited channel
                pendingRequest.deferred.onJoin {}
                exitStatus.onAwait {
                    throw it
                }
```

the current (working) code is:

```kotlin
launch(start = CoroutineStart.ATOMIC) {
         check(messageChannel.offer(clientMessage)) // unlimited channel
```

That's all for now, I will up to date this issue with further informations, when available.

**Comment by qwwdfsad** [comment]
Thanks for the details.
Could you please clarify, are you using `messageChannel` for send from multiple threads simultaneously? With `select` or with regular `send` as well?

**Comment by fvasco** [comment]
> Could you please clarify, are you using messageChannel for send from multiple threads simultaneously? 

Yes, we use the `Dispatchers.Default`, plus the mentioned `Undispatched` job.

> With select or with regular send as well?

Regural `send` _should_ not be called.

Channel is consumed by a single task

```kotlin
                        whileSelect {
                            messageChannel.onReceive { clientMessage ->
                                socket.send(clientMessage)
                                true
                            }
                            onTimeout(timeout) {
                                socket.send(Heartbeat)
                                true
                            }
                        }
```

In the above code `socket.send` is:

```kotlin
    override suspend fun send(message: Message) {
        sendMutex.withLock {
            sendBuffer.reset()
            encodeTo(message, sendBuffer)
            webSocket.sendText(sendBuffer, true)?.await() // JDK 11 WebSocket
        }
    }
```

**Comment by qwwdfsad** [comment]
We've recently fixed the bug in `onJoin` (#1130), maybe it could be the case.
Feel free to re-open the ticket if the problem is still here
